From 01137ef359ffafb3039bea556e7e32af2fc608f4 Mon Sep 17 00:00:00 2001
From: Markus Boehme <markus.boehme@mailbox.org>
Date: Fri, 14 Nov 2025 13:18:37 +0100
Subject: [PATCH] cloud-ssh-keys-fetcher: add initial policy (#8)

Add initial policy module for Chainguard's cloud-ssh-keys-fetcher. Drop
privileges from sshd_t by transitioning to a new less-privileged domain
that can essentially:

  1. Find out which cloud it is running on
2. Call the cloud-specific instance meta data service to fetch
acceptable public keys

Fixes: VMS-173
Fixes: VMS-174

Signed-off-by: Markus Boehme <markus.boehme@chainguard.dev>
---
 dist/targeted/modules.conf                    |  4 ++
 .../modules/contrib/cloud-ssh-keys-fetcher.fc |  1 +
 .../modules/contrib/cloud-ssh-keys-fetcher.if |  1 +
 .../modules/contrib/cloud-ssh-keys-fetcher.te | 43 +++++++++++++++++++
 4 files changed, 49 insertions(+)
 create mode 100644 policy/modules/contrib/cloud-ssh-keys-fetcher.fc
 create mode 100644 policy/modules/contrib/cloud-ssh-keys-fetcher.if
 create mode 100644 policy/modules/contrib/cloud-ssh-keys-fetcher.te

diff --git a/dist/targeted/modules.conf b/dist/targeted/modules.conf
index 1fc52c497..514380ade 100644
--- a/dist/targeted/modules.conf
+++ b/dist/targeted/modules.conf
@@ -680,6 +680,10 @@ clogd = module
 #
 cloudform = module
 
+# Layer: services
+# Module: cloud-ssh-keys-fetcher
+cloud-ssh-keys-fetcher = module
+
 # Layer: services
 # Module: cmirrord
 #
diff --git a/policy/modules/contrib/cloud-ssh-keys-fetcher.fc b/policy/modules/contrib/cloud-ssh-keys-fetcher.fc
new file mode 100644
index 000000000..c19efee0c
--- /dev/null
+++ b/policy/modules/contrib/cloud-ssh-keys-fetcher.fc
@@ -0,0 +1 @@
+/usr/bin/cloud-ssh-keys-fetcher -- gen_context(system_u:object_r:cloud_ssh_keys_fetcher_exec_t,s0)
diff --git a/policy/modules/contrib/cloud-ssh-keys-fetcher.if b/policy/modules/contrib/cloud-ssh-keys-fetcher.if
new file mode 100644
index 000000000..9eebb5c73
--- /dev/null
+++ b/policy/modules/contrib/cloud-ssh-keys-fetcher.if
@@ -0,0 +1 @@
+## <summary>Chainguard-provided sshd AuthorizedKeysCommand to fetch public keys from cloud meta data.</summary>
diff --git a/policy/modules/contrib/cloud-ssh-keys-fetcher.te b/policy/modules/contrib/cloud-ssh-keys-fetcher.te
new file mode 100644
index 000000000..73cc06db6
--- /dev/null
+++ b/policy/modules/contrib/cloud-ssh-keys-fetcher.te
@@ -0,0 +1,43 @@
+policy_module(cloud-ssh-keys-fetcher, 0.0.1)
+
+gen_require(`
+	type init_t;
+	type sshd_t;
+')
+
+########################################
+#
+# Declarations
+#
+
+type cloud_ssh_keys_fetcher_t;
+type cloud_ssh_keys_fetcher_exec_t;
+
+application_domain(cloud_ssh_keys_fetcher_t, cloud_ssh_keys_fetcher_exec_t)
+
+########################################
+#
+# Local policy
+#
+
+domain_auto_trans(sshd_t, cloud_ssh_keys_fetcher_exec_t, cloud_ssh_keys_fetcher_t)
+
+# cloud-ssh-keys-fetcher is a shell script invoking other binaries
+corecmd_exec_bin(cloud_ssh_keys_fetcher_t)
+
+# systemd-detect-virt: are we in a user namespace? check /proc/1/ns
+allow cloud_ssh_keys_fetcher_t init_t:dir search;
+fs_getattr_nsfs_files(cloud_ssh_keys_fetcher_t)
+
+# systemd-detect-virt: which cloud is this? check DMI product ID
+dev_read_sysfs(cloud_ssh_keys_fetcher_t)
+
+# curl: find home directory
+auth_read_passwd(cloud_ssh_keys_fetcher_t)
+
+# curl: read OpenSSL config and certs
+miscfiles_read_generic_certs(cloud_ssh_keys_fetcher_t)
+
+# curl: call cloud-specific meta-data services
+allow cloud_ssh_keys_fetcher_t self:tcp_socket create_socket_perms;
+corenet_tcp_connect_http_port(cloud_ssh_keys_fetcher_t)
-- 
2.51.2

